In [1]:
!pip install datasets
!pip install transformers
!pip install evaluate
import evaluate

from datasets import load_dataset
dataset = load_dataset("imdb")

from transformers import pipeline

bert_pipelineTrun = pipeline("sentiment-analysis", model="bert-base-uncased", truncation=True)
distilbert_pipelineTrun = pipeline("sentiment-analysis", model="distilbert-base-uncased-finetuned-sst-2-english", truncation=True)


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 480.6/480.6 kB 10.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 116.3/116.3 kB 10.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 179.3/179.3 kB 2.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 143.5/143.5 kB 7.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 194.8/194.8 kB 9.5 MB/s eta 0:00:00
  Attempting uninstall: fsspec
    Found existing installation: fsspec 2024.10.0
    Uninstalling fsspec-2024.10.0:
      Successfully uninstalled fsspec-2024.10.0
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
gcsfs 2024.10.0 requires fsspec==2024.10.0, but you have fsspec 2024.9.0 which is incompatible.
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 84.0/84.0 kB 5.0 MB/s eta 0:00:00


/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


README.md:   0%|          | 0.00/7.81k [00:00<?, ?B/s]

train-00000-of-00001.parquet:   0%|          | 0.00/21.0M [00:00<?, ?B/s]

test-00000-of-00001.parquet:   0%|          | 0.00/20.5M [00:00<?, ?B/s]

unsupervised-00000-of-00001.parquet:   0%|          | 0.00/42.0M [00:00<?, ?B/s]

Generating train split:   0%|          | 0/25000 [00:00<?, ? examples/s]

Generating test split:   0%|          | 0/25000 [00:00<?, ? examples/s]

Generating unsupervised split:   0%|          | 0/50000 [00:00<?, ? examples/s]

config.json:   0%|          | 0.00/570 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/440M [00:00<?, ?B/s]

Some weights of BertForSequenceClassification were not initialized from the model checkpoint at bert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

Device set to use cuda:0


config.json:   0%|          | 0.00/629 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/268M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

Device set to use cuda:0


In [3]:
texts = dataset["test"]["text"]
true_labels = dataset["test"]["label"]

bertResults = bert_pipelineTrun(texts)
distilbertResults = distilbert_pipelineTrun(texts)


In [8]:
bertLabels = [result["label"] for result in bertResults]
distilbertLabels = [result["label"] for result in distilbertResults]

bertlabel_mapping = {"LABEL_0": 0, "LABEL_1": 1}
distilbertlabel_mapping = {"NEGATIVE": 0, "POSITIVE": 1}
bertLabels = [bertlabel_mapping[label] for label in bertLabels]
distilbertLabels = [distilbertlabel_mapping[label] for label in distilbertLabels]

In [9]:
accuracy_metric = evaluate.load("accuracy")
f1_metric = evaluate.load("f1")

bertAccuracy = accuracy_metric.compute(predictions=bertLabels, references=true_labels)
bertF1 = f1_metric.compute(predictions=bertLabels, references=true_labels)

print("BERT Accuracy (Truncated):", bertAccuracy["accuracy"])
print("BERT F1 Score (Truncated):", bertF1["f1"])

distilbertAccuracy = accuracy_metric.compute(predictions=distilbertLabels, references=true_labels)
distilbertF1 = f1_metric.compute(predictions=distilbertLabels, references=true_labels)

print("DistilBERT Accuracy (Truncated):", distilbertAccuracy["accuracy"])
print("DistilBERT F1 Score (Truncated):", distilbertF1["f1"])

BERT Accuracy (Truncated): 0.47128
BERT F1 Score (Truncated): 0.17900621118012422
DistilBERT Accuracy (Truncated): 0.89072
DistilBERT F1 Score (Truncated): 0.8874794069192751




---


Models: "bert-base-uncased", "distilbert-base-uncased"

BERT Accuracy (Truncated): 0.50012

BERT F1 Score (Truncated): 0.6663106459106566

DistilBERT Accuracy (Truncated): 0.47568

DistilBERT F1 Score (Truncated): 0.6239384897865504



---


Models: "distilbert-base-uncased-finetuned-sst-2-english", "bert-base-uncased"


BERT Accuracy (Truncated): 0.47128

BERT F1 Score (Truncated): 0.17900621118012422

DistilBERT Accuracy (Truncated): 0.89072

DistilBERT F1 Score (Truncated): 0.8874794069192751



---


**Short Result Analysis:**

Bert had a higher accuracy and f1 score and overall performed better than distilBert. I thought the scores would be closer together but thinking back now I realize these two models went through different training and differing padding techniques. I currently have no idea on how to improve it other than training it specifically to the imdb reviews instead of other texts.



---


**Findings:**

I choose the IMDB dataset as I thought it was interesting as I quite enjoy movies and I think that movie reviews, especially the ones in the dataset would be hard for the pre-trained models to accurately get the sentiment.

*Note: I think there is something wrong with the the pre-trained model as reading the accuracy on HuggingFace it should be in the 90s but when I ran my code it came out to be around 50ish percent. That or I didn't implement my code correctly.



---

Note: I reran my code using a different model and distilBert did better so disregard everything previosly stated. I believe I chose the wrong model and it is leading to bad results. If I get the right Bert Model I will get the right value on accuracy.


---



